presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java

```diff
@@ -0,0 +1,389 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.pinot.PinotUtils.doWithRetries;
+import static com.google.common.base.Preconditions.checkState;
+
+public class PinotBrokerPageSource
+        implements ConnectorPageSource
+{
+    private static final String REQUEST_PAYLOAD_TEMPLATE = "{\"pql\" : \"%s\" }";
+    private static final String QUERY_URL_TEMPLATE = "http://%s/query";
+
+    private static final String PINOT_INFINITY = "∞";
+    private static final String PINOT_POSITIVE_INFINITY = "+" + PINOT_INFINITY;
+    private static final String PINOT_NEGATIVE_INFINITY = "-" + PINOT_INFINITY;
+
+    private static final Double PRESTO_INFINITY = Double.POSITIVE_INFINITY;
+    private static final Double PRESTO_NEGATIVE_INFINITY = Double.NEGATIVE_INFINITY;
+
+    private final PinotQueryGenerator.GeneratedPql brokerPql;
+    private final PinotConfig pinotConfig;
+    private final List<PinotColumnHandle> columnHandles;
+    private final PinotClusterInfoFetcher clusterInfoFetcher;
+    private final ConnectorSession session;
+    private final ObjectMapper objectMapper;
+
+    private boolean finished;
+    private long readTimeNanos;
+    private long completedBytes;
+
+    public PinotBrokerPageSource(PinotConfig pinotConfig, ConnectorSession session, PinotQueryGenerator.GeneratedPql brokerPql, List<PinotColumnHandle> columnHandles, PinotClusterInfoFetcher clusterInfoFetcher, ObjectMapper objectMapper)
+    {
+        this.pinotConfig = pinotConfig;
+        this.brokerPql = brokerPql;
+        this.clusterInfoFetcher = clusterInfoFetcher;
+        this.columnHandles = ImmutableList.copyOf(columnHandles);
+        this.session = session;
+        this.objectMapper = objectMapper;
+    }
+
+    private static Double parseDouble(String value)
+    {
+        try {
+            return Double.valueOf(value);
+        }
+        catch (NumberFormatException ne) {
+            switch (value) {
+                case PINOT_INFINITY:
+                case PINOT_POSITIVE_INFINITY:
+                    return PRESTO_INFINITY;
+                case PINOT_NEGATIVE_INFINITY:
+                    return PRESTO_NEGATIVE_INFINITY;
+            }
+            throw new PinotException(PinotErrorCode.PINOT_DECODE_ERROR, Optional.empty(), "Cannot decode double value from pinot " + value, ne);
+        }
+    }
+
+    private void setValue(Type type, BlockBuilder blockBuilder, String value)
+    {
+        if (value == null) {
+            blockBuilder.appendNull();
+            return;
+        }
+        boolean handled = true;
+        if (type instanceof FixedWidthType) {
```

Comment:
Doing these `instanceof` checks in Java is normally considered an anti-pattern. You want to find ways to avoid it best you can using interfaces and inheritance etc. In this case I'd consider doing leveraging function overloading to make this code have less conditionals.

Reply:
I agree but I need to switch on type here and this seemed like a simpler approach than creating many small functions and class to keep state to achieve overloading.

Reply:
Either way, let's get rid of the `handled` flag variable using the following refactor:
```
        if (value == null) {
            blockBuilder.appendNull();
            return;
        }
        if (!(type instanceof FixedWidthType) && !(type instanceof VarcharType)) {
            throw new PinotException(PinotErrorCode.PINOT_UNSUPPORTED_COLUMN_TYPE, Optional.empty(), "type '" + type + "' not supported");
        }

        if (type instanceof FixedWidthType) {
            completedBytes += ((FixedWidthType) type).getFixedSize();
            if (type instanceof BigintType) {
                type.writeLong(blockBuilder, parseDouble(value).longValue());
            }
            else if (type instanceof IntegerType) {
                blockBuilder.writeInt(parseDouble(value).intValue());
            }
            else if (type instanceof TinyintType) {
                blockBuilder.writeByte(parseDouble(value).byteValue());
            }
            else if (type instanceof SmallintType) {
                blockBuilder.writeShort(parseDouble(value).shortValue());
            }
            else if (type instanceof BooleanType) {
                type.writeBoolean(blockBuilder, Boolean.valueOf(value));
            }
            else if (type instanceof DecimalType || type instanceof DoubleType) {
                type.writeDouble(blockBuilder, parseDouble(value).doubleValue());
            }
            else if (type instanceof TimestampType) {
                type.writeLong(blockBuilder, Long.valueOf(value));
            }
            else {
                throw new PinotException(PinotErrorCode.PINOT_UNSUPPORTED_COLUMN_TYPE, Optional.empty(), "type '" + type + "' not supported");
            }
        }
        else { // Do not need to clarify that this is a Varchar type
            Slice slice = Slices.utf8Slice(value);
            blockBuilder.writeBytes(slice, 0, slice.length()).closeEntry();
            completedBytes += slice.length();
        }
```